File: sdk/nodejs/src/ext/extension-service.ts
import {ExtensionInfo} from "../extension-info";
import {Operator} from "./operator";
import {EntityInfo} from "../entity-info";
import {Handler} from "./handler";
import {Entity} from "../entity";
import {Event} from "../model";
import {Record} from "../model/record";

export interface ExtensionService {
    handler<T extends Entity>(entityInfo: EntityInfo): Handler<T>;

    run(): Promise<void>;

    close(): void;

    registerExtensionWithOperator(extensionInfo: ExtensionInfo, localOperator: (event: Event, entity: Record) => Record): string;

    unRegisterOperator(id: string): void;

    registerPendingItems(): Promise<void>;
}

File: sdk/java/src/main/java/io/apibrew/client/ext/ExtensionService.java
package io.apibrew.client.ext;

import io.apibrew.client.Entity;
import io.apibrew.client.EntityInfo;
import io.apibrew.client.ExtensionInfo;
import io.apibrew.client.model.Extension;
import io.apibrew.client.model.Record;

import java.io.IOException;
import java.util.function.BiFunction;

public interface ExtensionService extends AutoCloseable {
    <T extends Entity> Handler<T> handler(Class<T> userClass);
    <T extends Entity> Handler<T> handler(EntityInfo<T> entityInfo);

    GenericHandler genericHandler();

    void run() throws IOException;
    void runAsync() throws IOException;

    String registerExtensionWithOperator(ExtensionInfo extensionInfo, BiFunction<Extension.Event, Record, Record> operator);

    void unRegisterOperator(String id);

    void registerPendingItems();
}
